### DISCUSSION TITLE
How to deal with git sparse-checkouts and therefore partial installs

### DISCUSSION THREAD

**Discussion by tiagobento** [pnpm/pnpm#4721] [Q&A] [answered]
Hey everyone!

Recently, my team (https://github.com/kiegroup/kie-tools) migrated from Lerna/Yarn to pnpm. Mostly for installing dependencies, but in the future we want to expand its use and completely move away from Lerna/Yarn in favor of pnpm only or combined with other tools as well.

Anyway, we currently have a medium-sized monorepo (74 packages), and we expect it to grow, so we're already taking the necessary measures to make sure we won't suffer (too much).

One of the things I'm currently researching is the possibility of working with sparse-checkouts [1][2]. This would allow any monorepo developer to not pay the heavy price of cloning the entire monorepo, while working only on a subset of it.

The script below will clone and build a subset of the packages, respecting their dependencies and all.

```
git clone --filter=blob:none --no-checkout https://github.com/kiegroup/kie-tools
cd kie-tools
git sparse-checkout init --cone
git checkout main
git sparse-checkout set scripts packages/envelope packages/envelope-bus packages/build-env packages/run-script-if
ls packages
pnpm bootstrap -F @kie-tools-core/envelope...
pnpm -r -F @kie-tools-core/envelope... build:prod
pnpm pretty-quick
git status
```

If you run it, you'll see that as a result, `pnpm-lock.yaml` is changed (please ignore the changes on `packages/build-env/graph.dot`, this is something we can solve internally), since pnpm thinks that the number of packages changed, as so many of the packages are not there anymore.

This is not ideal, and I suspect that everyone working with a shared workspace file will be experiencing this issue if working with sparse-checkouts.

We're already exploring the possibility of moving away from the shared workspace lockfile (--shared-workspace-lockfile=false), which I think could solve this issue?

I'd like to know your thoughts and if someone has already experienced this issue before.

Thanks in advance! :)

---
References:
[1] https://github.blog/2020-01-17-bring-your-monorepo-down-to-size-with-sparse-checkout/
[2] https://git-scm.com/docs/git-sparse-checkout

**Comment by tiagobento** [comment]
Update: moving away from the shared workspace lockfile is not an option for us, since we don't have a centralized place to reuse dependency versions from.

Imagine that package A depends on `lodash`, and package B doesn't. When not using a shared workspace lockfile, if I install `lodash` on package B, the versions are potentially different, which may cause problems.

**Comment by tiagobento** [comment]
Bump?

**Comment by zkochan** [comment]
Can't you just ignore the changes that happen to the lockfile in this case?

**Comment by tiagobento** [comment]
If I don't change any dependencies on any packages, yes, but if I do, then I need to re-generate the lockfile to take my changes into account.

**Reply by zkochan** [reply]
Maybe it can be done if you use filter to specify all projects that you have checked out. Like

```
pnpm install -F foo -F bar
```

But it won't work with other commands. Like with `pnpm add`.

**Reply by tiagobento** [reply]
@zkochan It works! I thought I was already doing that, but my `bootstrap` script actually has another part to it, so the `-F @kie-tools-core/envelope...` argument was not getting picked up by the `pnpm install` call.

**Comment by tiagobento** [comment, marked as answer]
For future reference, this is what worked -> https://github.com/pnpm/pnpm/discussions/4721#discussioncomment-3109734

**Reply by tiagobento** [reply]
Just discovered that if you add `-F .` to the mix, the lockfile changes. Solution is to add `--frozen-lockfile` during install.

**Reply by tiagobento** [reply]
Updated script:

```bash
git clone --filter=blob:none --no-checkout --depth 1 https://github.com/kiegroup/kie-tools
cd kie-tools
git sparse-checkout init --cone
git checkout main
git sparse-checkout set scripts repo
echo "Discovering which packages to fetch for @kie-tools-core/envelope..."
git sparse-checkout set scripts repo packages/envelope packages/envelope-bus packages/build-env packages/run-script-if
ls -la packages
pnpm install-dependencies -F @kie-tools-core/envelope... -F . --frozen-lockfile && pnpm link-packages-with-self
pnpm -F @kie-tools-core/envelope... build:dev
pnpm pretty-quick
git status
```

**Comment by tiagobento** [comment]
`pnpm add` command works normally with the expected changes being done to `pnpm-lock.yaml`. i.e. not removing the packages that are not present in the sparse checkout.

`pnpm remove`, however, removed all packages that were not present in the sparse checkout. A workaround I found was manually editing the `package.json` files and running a filtered `pnpm install`, like `pnpm install -F @kie-tools-core/envelope...`.

With these steps, it's possible to add and remove dependencies from the packages present in a sparse checkout.
